docs/range2cidr.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" (C) Copyright 2020 James Cuzella <james.cuzella@lyraphase.com>,
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH RANGE2CIDR 1 "June  9 2020"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
range2cidr \- program to convert iblocklist IP ranges to CIDR format
.SH SYNOPSIS
.B range2cidr
.RI [ options ] " IP-RANGE" ...
.br
.B range2cidr
.RI [ options ] " < files" ...
.br
cat files... |
.B range2cidr
.SH DESCRIPTION
This manual page documents briefly the
.B range2cidr
command.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
\fBrange2cidr\fP is a program that converts iblocklist / bluetack style
IP ranges to CIDR format.
.SH INTRODUCTION
This program takes input as IP ranges with dash separated (`-') pairs,
and converts these into their equivalent CIDR range notation(s).

For example:
.TP
 IP-RANGE
.RS 10
 => CIDR(s)
.RE
.TP
 1.22.65.175-1.22.65.175
.RS 10
 => 1.22.65.175/32
.RE
.TP
 192.168.0.0-192.168.255.255
.RS 10
 => 192.168.0.0/16
.RE
.TP
 86.66.128.0-86.70.225.255
.RS 10
 => 86.66.128.0/17
    86.67.0.0/16
    86.68.0.0/15
    86.70.0.0/17
    86.70.128.0/18
    86.70.192.0/19
    86.70.224.0/23
.RE
.SH OPTIONS
This program follows the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the Info files.
.PP
.I range2cidr
accepts the following options.
.TP
.PD 0
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-V, \-\-version
Show version of program.
.TP
.BI \-f " filename"
.TP
.PD
.BI \-\^\-file " filename"
.TP
.PD
.BI \-\^\-file "=filename"
Read the IP range input from the file
.IR filename ,
instead of from the first command line argument.
.br
Multiple
.B \-f
(or
.BR \-\^\-file )
options may be used.
.TP
.PD 0
.BI \-?
.TP
.PD
.BI \-\^\-help
.TP
.PD
.BI \-\^\-usage
Display brief usage message
.SH SEE ALSO
.BR ipcalc (1),
.BR sipcalc (1).
.SH AUTHOR
.br
Michele Santullo
.RS 4
range2cidr creator.
.RE
James Cuzella
.RS 4
modifications to support command line flags, autotools integration, and Debian packaging.
.RE
.SH COPYRIGHT
.br
Copyright © 2017 Michele Santullo
Copyright © 2020 James Cuzella
.RE
.br
This manual page was written by James Cuzella <james.cuzella@lyraphase.com>, for the Debian project (and may be used by others).



CHANGELOG.md

<a name="v0.0.6"></a>
## [v0.0.6](https://github.com/trinitronx/range2cidr/compare/distribution/range2cidr-0.0.5...v0.0.6)

> 2020-06-13

### Bug Fixes

* **git-chglog:** CHANGELOG GoTemplate now supports commit messages without special "type(scope): Subject" format

### Features

* Adding HACKING.md docs about how to work on this repo
* **GitHub Actions:** Adding GitHub Release Workflow


<a name="distribution/range2cidr-0.0.5"></a>
## [distribution/range2cidr-0.0.5](https://github.com/trinitronx/range2cidr/compare/release/range2cidr-0.0.5...distribution/range2cidr-0.0.5)

> 2020-06-12

* Distribution of range2cidr version 0.0.5

<a name="release/range2cidr-0.0.5"></a>
## [release/range2cidr-0.0.5](https://github.com/trinitronx/range2cidr/compare/v0.0.5...release/range2cidr-0.0.5)

> 2020-06-12


<a name="v0.0.5"></a>
## [v0.0.5](https://github.com/trinitronx/range2cidr/compare/v0.0.4...v0.0.5)

> 2020-06-12

* .gitgnore all autotools generated files
* Adding extra clean steps to dist-hook for pristine tarball
* Add GPL-3+ copyright notice to autotools input files
* Link configure script revision to git-version-gen output
* Hook up git-dist Makefile.am include
* Fix GNU-ism in git-dist Makefile template
* Adding Roger Leigh's git-dist.mk Automake template from schroot project
* Refactor autotools generated config.h includes
* Ensure manpage is included in dist tarball
* Move manpage to man/ dir & tell autotools about it
* Ignore autom4te.cache files
* Un-ignore autotools generated files required for source distribution
* bootstrap.sh now calls autogen.sh - FULL initial autotools bootstrap
* Add internal verbose.h to noinst_HEADERS (so make dist tarball will compile!)
* Add format attribute to verbose / printf variadic function call

<a name="v0.0.4"></a>
## [v0.0.4](https://github.com/trinitronx/range2cidr/compare/v0.0.3...v0.0.4)

> 2020-06-09

* Add some missed files to clean during clean-local-am
* Update manpage & pull it into upstream branch
* Add --version (-V) option with GPL-3+ notice
* Add friendly configure output for Header & Library checks
* Reimplement argv processing with ptopt when HAVE_LIBPOPT defined
* Replace all printf with new verbose() function
* Add initial popt implementation of command line flag parsing
* Add clean-local-am autoreconf clean target
* Hook up AX_IS_RELEASE & AX_COMPILER_FLAGS to automake CXXFLAGS, LDFLAGS
* Vendor autoconf-archive AX_IS_RELEASE & AX_COMPILER_FLAGS M4 templates
* Only remove .*version files during maintainer-clean target
* Hook autoconf & automake into using git-version-gen + .*version files
* Switch to pulseaudio's git-version-gen script
* Add GNOME version of git-version-gen script

<a name="v0.0.3"></a>
## [v0.0.3](https://github.com/trinitronx/range2cidr/compare/v0.0.2...v0.0.3)

> 2020-06-08

* Add configure checks for boost library & headers

<a name="v0.0.2"></a>
## [v0.0.2](https://github.com/trinitronx/range2cidr/compare/v0.0.1...v0.0.2)

> 2020-06-08

* Embed COPYING & INSTALL files that automake wants into the repo
* Add autoconf / automake generated files to: make maintainer-clean target
* bootstrap.sh: Run automake --install-missing
* Vendor autoconf-archive boost M4 templates
* Switching to autoconf / automake
* Move source files to src/
* Get *FLAGS vars from dpkg-buildflags
* Makefile: List .PHONY targets

<a name="v0.0.1"></a>
## v0.0.1

> 2020-06-07

* Add README
* Fix space formatting
* Adding range2cidr initial code


src/range2cidr.cpp
#include <global.h>
#include <string>
#include <iostream>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <boost/utility/string_ref.hpp>
#include <vector>
#include <algorithm>
#include <array>
#include <cstdint>
#include <unistd.h>
#include <boost/range/algorithm/copy.hpp>
#include <ciso646>
#ifdef HAVE_POPT_H
#include <popt.h>
#endif
#include <stdbool.h>

PACKAGE_COPYRIGHT
PACKAGE_LICENSE

namespace {
	uint32_t ipv4_to_uint (boost::string_ref parIP) {
		const auto max_ip_str_len = 3 * 4 + 3 + 1;
		struct in_addr addr;
 
		std::array<char, max_ip_str_len> ip_cpy;
		std::copy(parIP.begin(), parIP.end(), ip_cpy.begin());
		ip_cpy[parIP.size()] = '\0';
		inet_aton(ip_cpy.data(), &addr);
		return ntohl(addr.s_addr);
	}
 
	//see: http://stackoverflow.com/questions/28339094/calculating-list-of-subnets-that-make-a-given-ip-range
	std::vector<std::string> subnets (uint32_t parStartIP, uint32_t parEndIP) {
		std::vector<std::string> retval;
		uint32_t host_bits = 0;
		uint32_t tmp = parStartIP ^ parEndIP;
		while(tmp != 0){
			tmp = tmp >> 1;
			host_bits++;
		}
		const uint32_t host_mask = static_cast<uint32_t>(-1) >> (32 - host_bits);
		uint32_t network_addr = parStartIP & (-1 ^ host_mask);
		uint32_t broadcast_addr = parStartIP | host_mask;
		if(host_bits > 1){
			const uint32_t split_low = (network_addr | host_mask >> 1);
			const uint32_t split_high =(broadcast_addr & (-1 ^ host_mask >> 1));
			if(parStartIP != network_addr || parEndIP != broadcast_addr){
				retval = subnets(parStartIP, split_low);
				auto temp_vec = subnets(split_high, parEndIP);
				retval.insert(retval.end(), std::make_move_iterator(temp_vec.begin()), std::make_move_iterator(temp_vec.end()));
			}
			else {
				struct in_addr ip_addr;
				ip_addr.s_addr = htonl(parStartIP);
				retval.push_back(std::string(inet_ntoa(ip_addr)) + "/" + std::to_string(32-host_bits));
			}
		}
		else {
			struct in_addr ip_addr;
			ip_addr.s_addr = htonl(parStartIP);
			retval.push_back(std::string(inet_ntoa(ip_addr)) + "/" + std::to_string(32-host_bits));
		}
		return retval;
	}
 
	std::vector<std::string> range_to_cidr (boost::string_ref parFrom, boost::string_ref parTo) {
		const auto from = ipv4_to_uint(parFrom);
		const auto to = ipv4_to_uint(parTo);
 
		return subnets(from, to);
	}
 
	void print_range (boost::string_ref parRange) {
		const auto hyphen_pos = parRange.find('-');
		if (boost::string_ref::npos == hyphen_pos) {
			std::cout << parRange << std::endl;
		}
		else {
			boost::range::copy(
				range_to_cidr(parRange.substr(0, hyphen_pos), parRange.substr(hyphen_pos + 1)),
				std::ostream_iterator<std::string>(std::cout, "\n")
			);
		}
	}
} //unnamed namespace
 


int main (int argc, char* argv[]) {
#ifdef HAVE_LIBPOPT
    verbose("(%s:%d) whee...we have popt.h\n",__FILE__,__LINE__);
    /* option parsing variables */
    char ch;
    poptContext opt_con;   /* context for parsing command-line options */
    char *extra_arg;
    static const char *f="";

    static struct poptOption options_table[] = {
        { "file", 'f', POPT_ARG_STRING, &f, 'f', "read IP range input from file", "STRING" },
        { "verbose", 'v', POPT_ARG_NONE, NULL, 'v', "enable verbose", "" },
        { "version", 'V', POPT_ARG_NONE, NULL, 'V', "display version", "" },
        POPT_AUTOHELP
        { NULL, 0, 0, NULL, 0 } /* end-of-list terminator */
    };

    opt_con = poptGetContext(NULL, argc, (const char **)argv, options_table, 0);

    verbose("Verbose is off\n");
    /* Now do options processing */
    while ((ch = poptGetNextOpt(opt_con)) >= 0) {
        verbose("between while & switch: ch = %c\n", ch);
        switch (ch) {
            case 'f':
                    verbose("handling 'f' option.\n");
                    break;
            case 'v':
                    verbose("handling 'v' option.\n");
                    setVerbose(true);
                    verbose("Verbose is on\n");
                    break;
            case 'V':
                    verbose("handling 'V' option.\n");
                    printf("%s\n\t%s\n\n%s",PACKAGE_STRING,COPYRIGHT,LICENSE);
                    exit(EXIT_SUCCESS);
                    break;
            default:
                    printf("unknown option '%c'.\n",ch);
                    exit(EXIT_FAILURE);
                    break;
        }
    }

    if (ch < -1) {
        // the user specified an invalid option, tell them
        poptPrintHelp(opt_con, stderr, 0);
        exit(EXIT_FAILURE);
    }


    verbose("(%s:%d) s = '%s'\n",__FILE__,__LINE__,f);
    verbose("(%s:%d) v = %d\n",__FILE__,__LINE__,getVerbose());
#else
	verbose("(%s:%d) rats...we don't have popt.h\n",__FILE__,__LINE__);
#endif

	using std::string;
	using boost::string_ref;
 
	if (not isatty(STDIN_FILENO)) {
		string curr_range;
		while (std::cin >> curr_range) {
			print_range(curr_range);
		}
	}
	else {
#if HAVE_LIBPOPT
    /* non-option args */
    while ((extra_arg = (char *)poptGetArg(opt_con))) {
        verbose("extra arg: %s\n", extra_arg);
        print_range(string_ref(extra_arg));
    }


    /* cleanup */
    poptFreeContext(opt_con);
#else
        for (int z = 1; z < argc; ++z) {
            verbose("argc:    %d\n",argc);
            verbose("argv:    %s\n",*argv);
            verbose("argv[z]: %s\n",argv[z]);
            print_range(string_ref(argv[z]));
        }
#endif
    }
 
	return 0;
}
